File: 1. Arrays And String/Question-3/Solution.py
#!/usr/bin/env python3

def urlify_list(string, true_length):
    """
    Replace spaces with '%20' using character array approach (simulating in-place)
    Time Complexity: O(n)
    Space Complexity: O(n) for the list conversion
    """
    # Convert to list for easier manipulation
    char_list = list(string)
    
    # Count spaces
    space_count = 0
    for i in range(true_length):
        if char_list[i] == ' ':
            space_count += 1
    
    # Calculate new index position
    index = true_length + space_count * 2
    
    # If there is extra space in the array, add null terminator
    if true_length < len(char_list):
        char_list[true_length] = '\0'
    
    # Replace spaces from end to beginning
    for i in range(true_length - 1, -1, -1):
        if char_list[i] == ' ':
            char_list[index - 1] = '0'
            char_list[index - 2] = '2'
            char_list[index - 3] = '%'
            index -= 3
        else:
            char_list[index - 1] = char_list[i]
            index -= 1
    
    # Convert back to string and return
    return ''.join(char_list).replace('\0', '')

def urlify_pythonic(string, true_length):
    """
    Replace spaces with '%20' using Python's built-in methods
    Time Complexity: O(n)
    Space Complexity: O(n) for the new string
    """
    # Trim the string to the true length and replace spaces
    return string[:true_length].replace(' ', '%20')

if __name__ == "__main__":
    # Example with list-based approach
    input_str = "Mr John Smith    "
    true_length = 13  # Length of "Mr John Smith"
    
    print(f'Before: "{input_str}"')
    result1 = urlify_list(input_str, true_length)
    print(f'After (list approach): "{result1}"')
    
    # Example with Pythonic approach
    result2 = urlify_pythonic(input_str, true_length)
    print(f'After (Pythonic approach): "{result2}"')
    
    # Additional test cases
    test_cases = [
        ("much ado about nothing      ", 22),
        ("   hello world  ", 13),
        ("", 0),
        ("  ", 2)
    ]
    
    for test_str, length in test_cases:
        print(f'\nInput: "{test_str}" (true length: {length})')
        print(f'Output: "{urlify_pythonic(test_str, length)}"')

File: 1. Arrays And String/Question-3/Solution.cpp
#include <iostream>
#include <string>
#include <cstring>
using namespace std;

/**
 * Replaces spaces with '%20' in a string
 * Time Complexity: O(n) where n is the length of the string
 * Space Complexity: O(1) as we're modifying the string in-place
 */
void replaceSpaces(char* str, int trueLength) {
    // Count spaces
    int spaceCount = 0;
    for (int i = 0; i < trueLength; i++) {
        if (str[i] == ' ') {
            spaceCount++;
        }
    }
    
    // Calculate new length
    int newLength = trueLength + spaceCount * 2;
    
    // Set null terminator at the end
    str[newLength] = '\0';
    
    // Perform the replacement from end to beginning
    for (int i = trueLength - 1; i >= 0; i--) {
        if (str[i] == ' ') {
            str[newLength - 1] = '0';
            str[newLength - 2] = '2';
            str[newLength - 3] = '%';
            newLength -= 3;
        } else {
            str[newLength - 1] = str[i];
            newLength--;
        }
    }
}

// C++ string version (more convenient)
string urlify(string s, int trueLength) {
    // Count spaces
    int spaceCount = 0;
    for (int i = 0; i < trueLength; i++) {
        if (s[i] == ' ') {
            spaceCount++;
        }
    }
    
    // Resize string to fit all characters
    int newLength = trueLength + 2 * spaceCount;
    s.resize(newLength);
    
    // Replace spaces from end to beginning
    for (int i = trueLength - 1; i >= 0; i--) {
        if (s[i] == ' ') {
            s[newLength - 1] = '0';
            s[newLength - 2] = '2';
            s[newLength - 3] = '%';
            newLength -= 3;
        } else {
            s[newLength - 1] = s[i];
            newLength--;
        }
    }
    
    return s;
}

int main() {
    // C-style string example
    char str[256] = "Mr John Smith    ";
    int trueLength = 13; // Length of "Mr John Smith"
    
    cout << "Before (C-style): \"" << str << "\"" << endl;
    replaceSpaces(str, trueLength);
    cout << "After (C-style): \"" << str << "\"" << endl;
    
    // C++ string example
    string s = "Mr John Smith";
    cout << "Before (C++ style): \"" << s << "\"" << endl;
    string result = urlify(s, s.length());
    cout << "After (C++ style): \"" << result << "\"" << endl;
    
    return 0;
}
